src/core/handle.ml

```diff
@@ -196,116 +281,75 @@ let handle_cmd : sig_state -> p_command -> sig_state * proof_data option =
-        fatal cmd.pos "Property modifiers cannot be used in theorems.";
-      if mstrat <> Eager then
-        fatal cmd.pos "Pattern matching strategy modifiers cannot be used \
-                       in theorems.";
-      (* Scoping the type (statement) of the theorem. *)
-      let a = scope_basic expo a in
-      (* Check that [a] is typable and that its type is a sort. *)
-      Typing.sort_type [] a;
-      (* We check that no metavariable remains in [a]. *)
-      if Basics.has_metas true a then
-        begin
-          fatal_msg "The type of [%s] has unsolved metavariables.\n" x.elt;
-          fatal x.pos "We have %s : %a." x.elt pp_term a
-        end;
-      (* Initialize proof state and save configuration data. *)
-      let st = Proof.init x a in
-      Console.push_state ();
-      (* Build proof checking data. *)
-      let finalize ss st =
-        Console.pop_state ();
-        match pe.elt with
-        | P_proof_abort ->
-            (* Just ignore the command, with a warning. *)
-            wrn cmd.pos "Proof aborted."; ss
-        | P_proof_admit ->
-            (* If the proof is finished, display a warning. *)
-            if Proof.finished st then
-              wrn cmd.pos "The proof is finished. You can use 'qed' instead.";
-            (* Add a symbol corresponding to the proof, with a warning. *)
-            out 3 "(symb) %s (admit)\n" x.elt;
-            wrn cmd.pos "Proof admitted.";
-            Sig_state.add_symbol ss expo Const Eager x a impl None
-        | P_proof_qed   ->
-            (* Check that the proof is indeed finished. *)
-            if not (Proof.finished st) then
-              begin
-                let _ = Tactics.handle_tactic ss st (none P_tac_print) in
-                fatal cmd.pos "The proof is not finished."
-              end;
-            (* Add a symbol corresponding to the proof. *)
-            out 3 "(symb) %s (qed)\n" x.elt;
-            Sig_state.add_symbol ss expo Const Eager x a impl None
+      (* If a type [ao = Some a] is given, then we check that it is typable by
+         a sort and that [t] has type [a]. Otherwise, we try to infer the type
+         of [t]. Unification goals are collected *)
+      let sort_goals, a = Proof.goals_of_typ x.pos ao t in
+      (* And the main "type" goal *)
+      let proof_term = fresh_meta ~name:x.elt a 0 in
+      let typ_goal = Proof.goals_of_meta proof_term in
+      let goals = sort_goals @ typ_goal in
+      (* Proof script *)
+      let (ts,pe) =
+        match ts_pe,p_t with
+        | None,Some p_t ->
+            let refine = Pos.make cmd.pos (P_tac_refine(p_t)) in
+            let ts = [refine] in
+            let pe = Pos.make cmd.pos P_proof_end in
+            (ts,pe)
+        | Some(ts,pe),None -> (ts,pe)
+        | Some _, Some _ ->
+            fatal cmd.pos "Proof script and proof_term definition !?"
+        | None, None  ->
+            fatal cmd.pos "NO Proof script and NO proof_term definition !?"
```

Comment:
Why these fatal errors? If there is a definition, then the first tactic must be refine. Perhaps, we could instead do something like:
```
let (ts,pe) =
  let (ts,pe) =
    match ts_pe with
    | None -> ([], Pos.make cmd.pos P_proof_end)
    | Some(ts,pe) -> (ts,pe)
  in
  match p_t with
  | Some p_t ->
    let refine = Pos.make cmd.pos (P_tac_refine(p_t)) in
    (refine::ts, pe)
  | None -> (ts,pe)
```

Reply:
We discussed last time that the (proof + term definition) case should be removed (because we can have a term definition + a unification tactic)
But why not filtering this kind of definition without proof nor term definition
```
definition a : TYPE
```
(with your changes.... this will generate a goal and generate ultimately a ``The proof is not finished`` error)

Reply:
HI Rehan. What is the problem in this case? The user wants to define a type a but provides no definition: this is an error.

Reply:
yes it's an error... so are you ok with this ?

```ocaml
      if ts_pe = None && p_t = None then
        fatal cmd.pos "no proof script nor term definition";
      let (ts,pe) =
        let (ts,pe) =
          match ts_pe with
          | None -> ([], Pos.make cmd.pos P_proof_end)
          | Some(ts,pe) -> (ts,pe)
        in
        match p_t with
        | Some p_t ->
          let refine = Pos.make cmd.pos (P_tac_refine(p_t)) in
          (refine::ts, pe)
        | None -> (ts,pe)
      in
```

Reply:
Yes, in the case of a definition.